libs/auth/src/java_card.ts

```diff
@@ -0,0 +1,462 @@
+              error.statusWord()
+            ),
+        };
+      }
+      return { response: 'error' };
+    }
+    return { response: 'correct' };
+  }
+
+  program(): Promise<void> {
+    return Promise.resolve();
+  }
+
+  readData(): Promise<Buffer> {
+    return Promise.resolve(Buffer.from([]));
+  }
+
+  writeData(): Promise<void> {
+    return Promise.resolve();
+  }
+
+  clearData(): Promise<void> {
+    return Promise.resolve();
+  }
+
+  unprogram(): Promise<void> {
+    return Promise.resolve();
+  }
+
+  /**
+   * We wrap this.readUser() because it:
+   * 1. Intentionally throws errors if any verification fails
+   * 2. Can throw errors due to external actions like preemptively removing the card from the card
+   *    reader
+   */
+  private async safeReadUser(): Promise<Optional<User>> {
+    try {
+      return await this.readUser();
+    } catch {
+      return undefined;
+    }
+  }
+
+  /**
+   * Reads the user on the card, performing various forms of verification along the way. Throws an
+   * error if any verification fails.
+   */
+  private async readUser(): Promise<Optional<User>> {
+    await this.selectApplet();
+
+    // Verify that the card VotingWorks cert was signed by VotingWorks
+    const cardVxCert = await this.retrieveCert(CARD_VX_CERT.OBJECT_ID);
+    await openssl([
+      'verify',
+      '-CAfile',
+      this.vxCertAuthorityCertPath,
+      cardVxCert,
+    ]);
+
+    // Verify that the card VxAdmin cert was signed by VxAdmin
+    const cardVxAdminCert = await this.retrieveCert(
+      CARD_VX_ADMIN_CERT.OBJECT_ID
+    );
+    const vxAdminCertAuthorityCert = await this.retrieveCert(
+      VX_ADMIN_CERT_AUTHORITY_CERT.OBJECT_ID
+    );
+    await openssl([
+      'verify',
+      '-CAfile',
+      vxAdminCertAuthorityCert,
+      cardVxAdminCert,
+    ]);
+
+    // Verify that the VxAdmin cert authority cert was signed by VotingWorks
+    // TODO: Figure out how to sign the VxAdmin cert authority cert with the VotingWorks cert
+    // authority cert
+    await openssl([
+      'verify',
+      '-CAfile',
+      this.vxCertAuthorityCertPath,
+      vxAdminCertAuthorityCert,
+    ]);
+
+    // Verify that the card has a private key that corresponds to the public key in the card
+    // VotingWorks cert
+    await this.verifyCardPrivateKey(CARD_VX_CERT.PRIVATE_KEY_SLOT, cardVxCert);
+
+    const user = await this.parseUserDataFromCert(cardVxAdminCert);
+
+    /**
+     * If the user doesn't have a PIN:
+     * Verify that the card has a private key that corresponds to the public key in the card
+     * VxAdmin cert
+     *
+     * If the user does have a PIN:
+     * Perform this verification later in checkPin because operations with this private key are
+     * PIN-gated
+     */
+    if (user?.role === 'poll_worker') {
+      await this.verifyCardPrivateKey(
+        CARD_VX_ADMIN_CERT.PRIVATE_KEY_SLOT,
+        cardVxAdminCert,
+        DEFAULT_PIN
+      );
+    }
+
+    return user;
```

Comment:
That's a great callout. Once I add automated tests, we'll at least be able to prevent accidental regressions. But let me think some more about this

Reply:
So thinking some more, I'm not actually that worried about this because, one level up from the card API, the auth API ensures that we go through PIN checking before considering a user logged in:

https://github.com/votingworks/vxsuite/blob/1a8394363bf604cdad68878b7937222ee07405c7/libs/auth/src/dipped_smart_card_auth.ts#L195-L206